### ISSUE TITLE
Process Start fails with URL on .NET Core 3

### ISSUE THREAD

**Issue by clairernovotny** [dotnet/runtime#28005]
The following code works on .NET Framework but throws on .NET Core 3:

`Process.Start("https://github.com");`

```
System.ComponentModel.Win32Exception
  HResult=0x80004005
  Message=The system cannot find the file specified.
  Source=System.Diagnostics.Process
  StackTrace:
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start(String fileName)
   at NetCoreProcessStartBug.MainWindow.ButtonBase_OnClick(Object sender, RoutedEventArgs e) in c:\dev\NetCoreProcessStartBug\NetCoreProcessStartBug\MainWindow.xaml.cs:line 31
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.RouteItem.InvokeHandler(RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.EventRoute.InvokeHandlers(Object source, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
   at System.Windows.Controls.Primitives.ButtonBase.OnClick()
   at System.Windows.Controls.Button.OnClick()
   at System.Windows.Controls.Primitives.ButtonBase.OnMouseLeftButtonUp(MouseButtonEventArgs e)
   at System.Windows.UIElement.OnMouseLeftButtonUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.RouteItem.InvokeHandler(RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.EventRoute.ReInvokeHandlers(Object source, RoutedEventArgs args)
   at System.Windows.UIElement.ReRaiseEventAs(DependencyObject sender, RoutedEventArgs args, RoutedEvent newEvent)
   at System.Windows.UIElement.CrackMouseButtonEventAndReRaiseEvent(DependencyObject sender, MouseButtonEventArgs e)
   at System.Windows.UIElement.OnMouseUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.RouteItem.InvokeHandler(RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.EventRoute.InvokeHandlers(Object source, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs args, Boolean trusted)
   at System.Windows.Input.InputManager.ProcessStagingArea()
   at System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   at System.Windows.Input.InputProviderSite.ReportInput(InputReport inputReport)
   at System.Windows.Interop.HwndMouseInputProvider.ReportInput(IntPtr hwnd, InputMode mode, Int32 timestamp, RawMouseActions actions, Int32 x, Int32 y, Int32 wheel)
   at System.Windows.Interop.HwndMouseInputProvider.FilterMessage(IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.WrappedInvoke(Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at System.Windows.Threading.Dispatcher.Invoke(DispatcherPriority priority, Delegate method, Object arg)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.TranslateAndDispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.Run()
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run(Window window)
   at System.Windows.Application.Run()
   at NetCoreProcessStartBug.App.Main()
```
GitHub repro: https://github.com/onovotny/NetCoreProcessStartBug

**Comment by AraHaan** [comment]
I tried the following codes on 3.1:

```cs
var psi = new ProcessStartInfo
{
    FileName = _https,
    UseShellExecute = true,
    Verb = "open",
};
using (var proc = Process.Start(psi))
{
}
```
And:

```cs
using (var proc = new Process())
{
    proc.StartInfo.FileName = _https;
    proc.StartInfo.UseShellExecute = true;
    proc.StartInfo.Verb = "open";
    proc.Start();
}
```
(I still prefer the last one though.) Also where ``_https`` is ``https://testing.us.sysb.ai``

However both results in:

```cs
Win32Exception: No application is associated with the specified file for this operation.   at System.Diagnostics.Process.StartWithShellExecuteEx(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at WebApi1.Ssh.SshHandler.Url() in C:\Users\User\Desktop\github\WebApi1\Ssh\SshHandler.cs:line 87
   at WebApi1.Ssh.SshHandler.Start() in C:\Users\User\Desktop\github\WebApi1\Ssh\SshHandler.cs:line 45
   at WebApi1.Program.Main(String[] args) in C:\User\Laura\Desktop\github\WebApi1\Program.cs:line 46
```

However this being Windows 10 on the latest fast ring insider build It should at least open up edge as the default browser or firefox depending on what the user set it to if I am correct still.

Who else wants to test this on net core 3.1?

**Comment by caioproiete** [comment]
@AraHaan I tested on .NET Core 3.1 on Windows 10 and worked fine (i.e. opened my default browser (Chrome) without any errors)

```
C:\>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.101
 Commit:    b377529961

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.101\

Host (useful for support):
  Version: 3.1.1
  Commit:  a1388f194c
```

**Comment by Ixus** [comment]
I found in a post from 2016 that this code supposedly worked in other platforms:

```
if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
{
    Process.Start("xdg-open", url);
}
else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
{
    Process.Start("open", url);
}
```

**Comment by soul4soul** [comment]
> 
> 
> Could anybody confirm if this code works on all platforms?
> 
> ```
> ProcessStartInfo psi = new ProcessStartInfo
> {
>     FileName = url,
>     UseShellExecute = true
> };
> Process.Start (psi);
> ```
> 
> Thanks.-

This is working on Windows 10. I'm happy with that since I'm targeting WPF which is windows only anyway.

**Comment by danmosemsft** [comment]
@soul4soul yes, it should -- assuming we can find xdg-open or similar app as shown above
https://github.com/dotnet/runtime/issues/28005#issuecomment-558716791

With the passing of time, I do not think we plan to change this behavior. There does not seem evidence of lots of existing code that cannot be changed to add UseShellExecute=true and that launches URL's.

**Comment by dynamicons** [comment]
Hello, and Thank you for providing the workaround, first of all. It can be done in one line:

               string my_web_link = "https://dynamic-applications.org/downloads/startup-product-manager/";

		// start the external weblink in a system process (as an executable):
		Process.Start( new ProcessStartInfo { FileName = my_web_link, UseShellExecute = true } );

As i have found out today, there seems to be a side-issue with installing **.NET 5.0 Desktop Environment**. I have managed to provide a Setup.exe (.vdproj) where the correct .NET 5.0 Desktop (x86) Installer download will be triggered from.

However, when i wanted to provide an Appx for the Windows 10 Store, the Tester always reported it will not run.

It seems to be so that a .NET executable contains a small, automatic stub that will check the environment at the beginning. Then it pops up a small requester allowing you to download the required version, just in case. However, with .NET 5.0 you can click "yes" or "no" in that built-in exe .NET installer, and nothing happens.

Without having any insight, i would guess this strange behavior could find its reason right here. 
So the .NET Stub calls Process.Start() and that one fails silently. Nobody thought of porting that command to the next version.

In effect, the self-check for all standard **.NET 5.0 Desktop Applications** (.exe), will fail completely. So if the proper .NET Framework or Desktop Environment is missing, they'll fail with a useless message when clicking the .exe, nothing happens, and they give no hint what to install. This may affect very many users naturally (Windows Forms and all traditional Desktop Apps).

Maybe someone could fix this for the next release? - 
thank you so much

best regards


Martin Bernhardt
mbe@dynamic-applications.org

**Comment by clairernovotny** [comment]
Hi @dynamicons, I'm not sure I understand the scenario you're describing as broken?

What exactly is failing? There should be no reason for a setup.exe to install the Desktop Runtime on end-user machines; self-contained apps are recommended here. I know [NuGet Package](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer) is distributed in the Store as a self-contained app and has no trouble with using Process.Start to launch a URL with UseShellExecute set to true.

**Comment by dynamicons** [comment]
Hello Claire, 

there seems to be a small self-check or self-repair component that's built-in to every .NET program.exe. 

With .NET 4.0 Framework, whenever the .NET Framework is missing (e.g. uninstalled by user on System clean-up), the application will show up a small dialog that will allow you to download .NET Framework again. So it's really like a built-in self-repair function.

Here is a Screenshot of the .NET self-repair dialog that shows up from the .exe itself on missing installation.

https://dynamions.files.wordpress.com/2020/11/iotdeveloper.exe-built-in-.net-5.0-detector.png

This functionality seems defunct with .NET 5.0 Desktop Environment (or compiling for .NET 5.0 target in VS 2019.8 with .NET 5.0). So the new functionality is that the self-repair of a .exe shows up the same popup but then it won't start your browser and download the .NET installer anymore. 

So that's why i thought it could be the same reason - the old Process.Start() call failing silently here.

The reason that i prefer the traditional Setup Installer is that this way you have to provide and deploy .NET 5.0 Desktop Environment only once. So the Setup.exe and the Appx package for Windows 10 Store are really, really small (4-5 MByte).

The self-contained all-in-one Exe generator is an interesting idea, of course. But it leads to 100 MByte IoTdeveloper.exe in my case. So that's way out of bounds of what i would find acceptable (double size with 10 MByte would be ok). The whole .NET Desktop Environment Setup.exe is only 50 MBytes, by the way. So the self-contained exe seems not be well compressed.

I have tried with Trimming and ReadyToRun=false but still it's pretty much 100 MBytes, minimum.
With a DSL upload connection of 1 MBit/s here, it would take all day to upload my Setups to the Windows 10 Store.

Also the user experience is slow and there is hundreds of files created that don't really have a purpose or justification to exist in parallel to each other, per program installation, like that. 

That's why i really prefer to provide a small Setup.exe and then only once download and install .NET 5.0 Desktop, separately.

**Comment by danmosemsft** [comment]
@dynamicons could you please open a new issue -- probably in dotnet/sdk -- proposing this "missing installation" experience?  Note that Windows has special knowledge of .NET Framework (via mscoree.dll etc), but it does not of .NET Core.

By opening a new issue we can get it routed to the right place - this issue is closed.

**Comment by dynamicons** [comment]
Hello Dan,

thank you - yes, that seems a good idea. 
I have filed in Issue 45542 against .NET Core development.

https://github.com/dotnet/runtime/issues/45542

It has already been reproduced, however, Process.Start() seems not to be the issue here.
The Error message itself was misleading, instead it hides other internal messages, and the feature itself also won't work.

However, the idea to fix this seems to be good for everyone.
